File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <set>
using namespace std;

void Find(vector<string> words, string begin, string target, int count, vector<int>& counts, set<string> visited, string cur)
{
	if (cur == target)
	{
		counts.push_back(count);
		return;
	}

	auto s = visited.insert(cur);
	if (s.second == false) return;

	for (int i = 0; i < words.size(); i++)
	{
		auto it = visited.find(words[i]);
		if (it != visited.end()) continue;

		int same = 0;
		for (int j = 0; j < words[i].size(); j++)
		{
			if (words[i][j] == cur[j]) same++;
		}
		if (same == cur.size() - 1)
		{
			Find(words, begin, target, count+1, counts, visited, words[i]);
		}
	}

}

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
	vector<int> counts;
	set<string> visited;
	auto it = find(words.begin(), words.end(), target);
	if (it == words.end())
	{
		return 0;
	}

	Find(words, begin, target, 0, counts,visited,begin);

	if(counts.size() != 0) return *min_element(counts.begin(), counts.end());
    else return 0;
    
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cs
using System;
using System.Linq;
using System.Collections.Generic;
public class Solution {
    public int solution(string begin, string target, string[] words) {
        int answer = 0;
if (!words.Contains(target)) return 0;

Stack<string> stk = new Stack<string>();

answer = route(words, begin, target, ref stk);

if (answer == int.MaxValue) answer = 0;
        return answer;
    }
    
    int route(string[] words,string cur,string target,ref Stack<string> stk)
{
    int res = int.MaxValue;

    Console.WriteLine($"cur = {cur}, target = {target}");
    if (cur == target)
    {
        Console.WriteLine("test");
        return stk.Count();
    }

    foreach(var item in words)
    {
        if(dif(cur,item) && !stk.Contains(item))
        {
            stk.Push(item);
            res = Math.Min(res, route(words, item, target,ref stk));
            stk.Pop();
        }
    }
    return res;
}

bool dif(string a, string b)
{
    int res = 0;
    for(int i = 0; i<a.Length;i++)
    {
        if (a[i] != b[i]) res++;
    }
    return res == 1 ? true : false;
}
}